### PR TITLE
[KEYCLOAK-16455][Adapter - JavaScript] Check if Keycloak realm available before proceed to Keycloak instance initialization + propagate errors where possible

### PR DESCRIPTION
**Pull request by afedotov**
The motivation for this is in case of Keycloak server realm is not available due to network error or Keycloak server down, the Promise returned from Keycloak.init({ onLoad: 'check-sso' }) is never being settled. So when using this JavaScript Adapter as recommended in documentation/tutorials with Vue.js or [keycloak-anagular](https://github.com/mauriciovigolo/keycloak-angular), the application is never being initialized and displays empty page.

### CODE REVIEW

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (mhajas) [inline adapters/oidc/js/src/main/resources/keycloak.js:1308]: Please use `createPromise()` function similarly as in other functions.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (mhajas) [inline adapters/oidc/js/src/main/resources/keycloak.js:1317]: Using Error object looks good to me and correct. However, we are not using it anywhere else. I think there may be some existing code that can be broken when we introduce this as it can expect errorData directly. Maybe we should not use Error now, but introduce enhancement Jira and replace it everywhere. WDYT?

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (mhajas) [inline adapters/oidc/js/src/main/resources/keycloak.js:1324]: It seems that this callback is not called when realm doesn't exist. Keycloak returns 404 response which doesn't pass `res.realm != kc.realm` check. So the error message in the catch(error) is `'Invalid response from Keycloak realm'` which doesn't seem correct.

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you for the PR @afedotov. I like the idea, however I added some comments to discuss. Also we would need some test coverage for this functionality. Here is some example of a test for this.

https://github.com/mhajas/keycloak/commit/7eb3fd84c6aed3be92256e6fc9d42d8705ce4333

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (afedotov) [reply adapters/oidc/js/src/main/resources/keycloak.js:1308]: Ok, let it be, no problem. As stated in source code, KeycloakPromise is deprecated in favor to native Promise, so this abstraction looks useless for me and I used native.
BTW, I am surprised why there is no universal XHR function, and XHR is inlined in every place, so it's hard to introduce some common behavior for XHR, for example it is very necessary option to configure network connection timeouts for XHR.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (afedotov) [reply adapters/oidc/js/src/main/resources/keycloak.js:1317]: You right. Error is canonical way to represent errors in JS, but as declared in TypeScript API (keycloak.d.ts), in case of error we should return KeycloakError. So lets respect this behavior.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (afedotov) [reply adapters/oidc/js/src/main/resources/keycloak.js:1324]: Yes, **onerror** handler is called only on fatal network errors. If any http response comes from server, then **onload** is called. We should do some fine grained http response code checking in **onload** handler.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (mhajas) [reply adapters/oidc/js/src/main/resources/keycloak.js:1308]: Yes you are right KeycloakPromise is deprecated, but deprecated doesn't mean removed yet. This way we can switch to native promises with a change just in one place in the future.

I agree with you. What you described sounds pretty good, I would be really happy to see this in the JS adapter in the future. If you would like to contribute this, feel free to create an enhancement Jira and provide a PR.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (mhajas) [reply adapters/oidc/js/src/main/resources/keycloak.js:1324]: Thank you for the explanation. In that case it seems that error messages on lines 1317 and 1324 should be interchanged. Because **onerror** means there was some invalid response not that realm is not available.

**Comment by afedotov** [comment]
Hello, @mhajas ! I'm just updated this PR as we discussed before. But there is problem with fine-grained error handling. Keycloak server does not set CORS headers on error responses, so any HTTP error response received from server is blocked by browser and XHR is always firing **onerror** callback instead of **onload**. So it is not possible to get HTTP error code or JSON error response body.

#### Existing realm, response contains Access-Control-Allow-Origin header:

```
$ curl -i -X GET -H "Accept: application/json" -H "Origin: http://localhost" https://sso.redhat.com/auth/realms/redhat-external/

HTTP/1.1 200 OK
Cache-Control: no-cache
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-Site: phx2
Content-Type: application/json
Content-Length: 973
Date: Thu, 15 Apr 2021 09:36:21 GMT
Connection: keep-alive
Set-Cookie: BIGipServer~prod~keycloak-webssl-https=879166730.64288.0000; path=/; Httponly; Secure
```

#### Invalid realm, no CORS headers in response:

```
$ curl -i -X GET -H "Accept: application/json" -H "Origin: http://localhost" https://sso.redhat.com/auth/realms/foobar/

HTTP/1.1 404 Not Found
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-Site: phx2
Content-Type: application/json
Content-Length: 32
Date: Thu, 15 Apr 2021 09:41:27 GMT
Connection: keep-alive
Set-Cookie: BIGipServer~prod~keycloak-webssl-https=862389514.64288.0000; path=/; Httponly; Secure

{"error":"Realm does not exist"}
```

Looks like it is not possible to do fine-grained error handling in XHR until Keycloak server will return CORS headers in error responses. Maybe we should open an enhancement ticket in JIRA on this?

**Comment by afedotov** [comment]
Hi, @mhajas !
Just created issue about CORS in realm resource error response:
https://issues.redhat.com/browse/KEYCLOAK-17799
When it will be resolved, **checkIfRealmAvailable()** be able to perform fine-grained error handling.
Without CORS there are only two outcomes: ok / unexpected error. No deep error details, but is usable somehow.

It remains to make test coverage. I'm need a bit of time to figure out Keycloak testsuite and then create tests.

**Comment by mhajas** [comment]
I remember I've already came across this issue: https://issues.redhat.com/browse/KEYCLOAK-12686 which was closed as `Won't do`. So I am not sure we will have the fine-grained error handling. On the other hand, it works even this way, so the better error handling is not must-have. 

Sure, take your time with the tests! Feel free to let me know should you need any help.

**Comment by afedotov** [comment]
Hello, @mhajas!
I've just added test cases and squash all changes to single commit.
Not sure about test that should simulate network failure when accessing Authorization Server. Decided to use URL with some random port that should be unused - https://localhost:12345/auth. Hope this approach would be ok in testsuite runtime environment.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:857]:
```suggestion
                .init(initOptions, assertErrorResponse("Keycloak server is not reachable, most likely due to network problems"));
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:839]: We should check also error response here:
```suggestion
                .init(initOptions, assertErrorResponse("Realm does not exist"));
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/AbstractJavascriptTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/AbstractJavascriptTest.java:219]: This can be replaced by this to work with comments in `JavascriptAdapterTest`.

```
    public void assertErrorResponse(String expectedError, WebDriver drv, Object output, WebElement evt) {
        Assert.assertNotNull("Empty error response", output);
        Assert.assertTrue("Invalid error response type", output instanceof Map);
        assertThat((Map<String, String>) output, anyOf(hasEntry("error", expectedError), hasEntry("error_description", expectedError)));
    }
    
    public JavascriptStateValidator assertErrorResponse(String expectedError) {
        return buildFunction(this::assertErrorResponse, expectedError);
    }
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:849]: This is not very nice :-(. But I can't think of any better solution. If nobody has a better idea we can leave this as is for now, and investigate if there are any failures in the future.

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you very much @afedotov! This look great, I would add just a little improvement into the testsuite and I believe we are good to go. 

@jonkoops you are quite involved in the Keycloak community around JS adapter, do you think you could have a quick look whether this breaks something for you?

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (jonkoops) [inline adapters/oidc/js/src/main/resources/keycloak.js:1779]: Instead of making the error that is returned here a Object I'd much rather see specific instances of the `Error` class. I know that in some other places in the code this is already done, but I would consider it a bad practice.

For example:
```ts
promise.setError(new Error('Keycloak server is not reachable, most likely due to network problems.'));
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by jonkoops** [comment]
@mhajas Looking at the code I don't see anything in particular that would trip things up, the additional error information is one thing that is really missing right now when handling exceptions from the adapter so this is a good improvement.

I am however a tad skeptical about the need to make a request just to see if a realm is available before embedding an iframe. @afedotov are you sure we cannot somehow capture this in the `processInit()` method when adding the iframes?

**Comment by jonkoops** [comment]
Note there is a similar PR that attempts to handle errors from iframes https://github.com/keycloak/keycloak/pull/7779. Unfortunately I have not had time to take a look at the issues there, but it might help with here.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (afedotov) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:857]: I did this intentionally, because it will success only inside this integration test, not in real world, due to different CORS environment. When running JavaScript test suite together with  default PhantomJS browser, CORS is allowed with PhantomJS' `--web-security=false` option. So XHR is not blocked on 404 HTTP error response. The same behavior I observe when running tests with Firefox browser. Not sure why CORS is allowed in this case, maybe because JS test code and local Authorization Server runs under same origin...

Commands I've used to run tests:
```
$ mvn -s maven-settings.xml -f testsuite/integration-arquillian/tests/base/pom.xml clean install -Dtest=JavascriptAdapterTest#checkInitWithInvalidRealm+checkInitWithUnavailableAuthServer -Pauth-server-wildfly

$ mvn -s maven-settings.xml -f testsuite/integration-arquillian/tests/base/pom.xml clean install -Dtest=JavascriptAdapterTest#checkInitWithInvalidRealm+checkInitWithUnavailableAuthServer -Pauth-server-wildfly -Dbrowser=firefox -Djs.browser=firefox -DfirefoxHeadless=true -Dwebdriver.gecko.driver=$HOME/.local/bin/geckodriver
```

So I've decided not to check error details and error message in tests, only fact that KeycloakError is returned from Promise, because this fine-grained details will not be available nowadays in real world scenario where errors are returned by Authorization Server without CORS headers and rejected by XHR.

So as you wish, we can make this changes and check error message contents in tests.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (afedotov) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java:849]: This is test environment/infrastructure task to provide **reliable** TCP service that can simulate TCP SYN-RST / ICMP Destination Unreachable / High latency and other network failures.

File: adapters/oidc/js/src/main/resources/keycloak.js
Comment (afedotov) [reply adapters/oidc/js/src/main/resources/keycloak.js:1779]: Hi, @jonkoops !
We can't use canonical JS Error objects now in JS Adapter because of:
- It will break contract specified in [keycloak.d.ts](https://github.com/keycloak/keycloak/blob/master/adapters/oidc/js/src/main/resources/keycloak.d.ts) TypeScript API, that states that in case of error KeycloakError should be returned
- @mhajas told that there is no place for now in JS Adapter where native Error objects used, and feared about compatibility issues.

Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/keycloak.js:1779]: Ok, fair enough. Let's keep it like this then. I'll see if we can make this a breaking change at some point.

**Comment by afedotov** [comment]
Hi, @jonkoops !
> are you sure we cannot somehow capture this in the `processInit()` method when adding the iframes?

The problem itself is in the `check3pCookiesSupported()` function that stands first in the domReady Promise chain:
```javascript
domReady().then(check3pCookiesSupported).then(processInit)
.catch(function() {
    promise.setError();
});
```
The `check3pCookiesSupported` callback is invoked from iframe that loaded from Authorization Server.
So in case of Authorization Server is unavailable due to some kind of server or network problems, `check3pCookiesSupported` callback is never being called and promise chain is stuck at the beginning.
In #7779 they try to fix this function trying to handle iframe errors, but it is not possible, and they stuck.
The only way in case of Authorization Server unavailability is to fail-fast **before** calling `check3pCookiesSupported()`, so this PR adds `checkIfRealmAvailable()` function in chain.

**Comment by jonkoops** [comment]
I think overall it's worth the overhead to get better error handing, so LGTM :heavy_check_mark:

**Review by jonkoops — DISMISSED** [review]

**Comment by afedotov** [comment]
Hi, @mhajas !
I've updated test cases as you suggested. I think it's ok to check error message contents for now,
hoping that KEYCLOAK-17799 issue will be fixed in future.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java]: I see, ok I didn't know that, thanks. In that case, we can leave it as is.

**Comment by mhajas** [comment]
@jonkoops  Thank you very much for the review!

@afedotov, I am not sure we can check the error response if it may differ on different browsers/environments. I will try to run more CI tests to check if it is causing problems. Please leave this as is now, I will let you know whether we need to revert or it is ok.

**Comment by afedotov** [comment]
> @afedotov, I am not sure we can check the error response if it may differ on different browsers/environments. I will try to run more CI tests to check if it is causing problems. Please leave this as is now, I will let you know whether we need to revert or it is ok.

Ok, @mhajas !

BTW, I'm trying to investigate and fix [KEYCLOAK-17799](https://issues.redhat.com/browse/KEYCLOAK-17799), and already have some progress. There is some challenge with existing test cases around this. Especially that in substance REST/JSON endpoints are tested using Selenium WebDriver and error assertion is done using resulting **HTML** error page validation :)

**Comment by mhajas** [comment]
@afedotov So we need to revert the last change you did with error checking. In the internal CI we have got following error:
```
Wrong error response
Expected: (map containing ["error"->"Realm does not exist"] or map containing ["error_description"->"Realm does not exist"])
     but: was <{error=network_error, error_description=Keycloak server is not reachable, most likely due to network problems}>
```

**Comment by afedotov** [comment]
> @afedotov So we need to revert the last change you did with error checking.

Done!

**Review by mhajas — APPROVED** [review]
Thank you very much @afedotov! Nice contribution.

**Review by mhajas — COMMENTED** [review]
@afedotov Hello, I discussed this again with others from the team and we realized this solution might be causing problems when somebody is using the JS adapter with another IDP than Keycloak.

Also, the extra request doesn't sound so bad, but in a functional production environment means mostly useless overhead. I plan to have a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race function. Using this function, we can fail the promise when the `check3pCookiesSupported` iframe doesn't return a message within 5 seconds for example. WDYT about this approach?

**Comment by afedotov** [comment]
Hello, @mhajas !

> might be causing problems when somebody is using the JS adapter with another IDP than Keycloak.

It will be very weird for me if someone comes up with idea to use **Keycloak** JavaScript adapter with another IDP. Is there any guarantees  stated in docs that it can be used as common OIDC adapter? Moreover there are already many places in this adapter that Keycloak-specific, and won't be working with another IDP. For example our notorious `check3pCookiesSupported()` procedure. Yes, it can be disabled by adapter settings, but nothing prevents for us to do the same for 'checkIfRealmAvailable()'. We can enable this "preflight" request to check if server is actually online only when `check3pCookiesSupported()` also enabled.
 
> Using this function, we can fail the promise when the `check3pCookiesSupported` iframe doesn't return a message within 5 seconds for example. WDYT about this approach?

Yes, Promise.race() is fine for this "timeout" approach, but it seems redundant.
We can patch `check3pCookiesSupported()` like this:

```diff
diff --git a/keycloak.js.orig b/keycloak.js
index 96d855b..3eba956 100644
--- a/keycloak.js.orig
+++ b/keycloak.js
@@ -1311,6 +1311,10 @@
                 iframe.style.display = 'none';
                 document.body.appendChild(iframe);
 
+                var handle = setTimeout(function () {
+                    promise.setError();
+                }, 5000);
+
                 var messageCallback = function(event) {
                     if (iframe.contentWindow !== event.source) {
                         return;
@@ -1327,6 +1331,7 @@
                             "silent check-sso are not available.")
                     }
 
+                    clearTimeout(handle);
                     document.body.removeChild(iframe);
                     window.removeEventListener("message", messageCallback);
                     promise.setSuccess();

```

Furthermore this can be made more versatile. We can extend `createPromise()` with optional `timeout` parameter.
So if timeout provided, newly created KeycloakPromise will be automatically rejected if not being fulfilled after timeout.

But "timeouts" are evil, especially in such "network" case. I would prefer explicit events.
What timeout value will be ok? For local networks? For internet? For mobile network?
It will be hard to balance between local and mobile network:
- if timeout it too short, slow clients on mobile networks will get false errors.
- if timeout is to long, this affects UX on local networks.

**Comment by afedotov** [comment]
Hello, @mhajas !

This is how `createPromise()` can be implemented to support optional timeout.

```javascript
        /**
         * Create new KeycloakPromise object that represents the eventual completion (or failure)
         * of an asynchronous operation and its resulting value.
         * 
         * Supports for optional settle timeout. If `promiseSettleTimeout` argument provided,
         * this promise will be automatically rejected unless not being settled before timeout.
         * 
         * @param {number} [promiseSettleTimeout] Optional promise settle timeout in milliseconds
         * @returns {KeycloakPromise} An eventual completion or failure
         */
        function createPromise(promiseSettleTimeout) {

            var timeoutHandle = null;

            var p = {
                setSuccess: function (result) {
                    clearTimeout(timeoutHandle);
                    p.resolve(result);
                },
                setError: function (result) {
                    clearTimeout(timeoutHandle);
                    p.reject(result);
                }
            };

            p.promise = new Promise(function (resolve, reject) {
                p.resolve = resolve;
                p.reject = reject;
            });

            p.promise.success = function (callback) {
                logPromiseDeprecation();
                this.then(function handleSuccess(value) {
                    callback(value);
                });
                return this;
            }

            p.promise.error = function (callback) {
                logPromiseDeprecation();
                this.catch(function handleError(error) {
                    callback(error);
                });
                return this;
            }

            if (promiseSettleTimeout > 0) {
                timeoutHandle = setTimeout(function () {
                    p.setError({ "error": "Promise is not settled within timeout of " + promiseSettleTimeout + "ms" });
                }, promiseSettleTimeout);
            }

            return p;

        }
```

**Comment by jonkoops** [comment]
I would not put implementation details of timeouts into the `createPromise` method. We want to get rid of it long-term.

**Comment by afedotov** [comment]
Hello, @jonkoops !

> I would not put implementation details of timeouts into the `createPromise` method. We want to get rid of it long-term.

I see nothing wrong with `createPromise()` function existing unless it is somehow exposed outside adapter API. It is commonly used Factory pattern. I guess that main goal is that Javascript Adapter **Public API** should return eventual completion results compatible with native Promise/Thenables, so they can be used with ES2017 `async / await`.

Anyway, if in long-term we want to use native Promise/Thenables in common with async/await inside private adapter code, then we can go like this:

```javascript
// Function to extend existing native Promise with timeout
function applyTimeoutToPromise(promise, timeout) {
    var timeoutHandle = null;
    var timeoutPromise = new Promise(function (resolve, reject) {
        timeoutHandle = setTimeout(function () {
            reject({ "error": "Promise is not settled within timeout of " + timeout + "ms" });
        }, timeout);
    });
    return Promise.race([promise, timeoutPromise]).finally(function () {
        clearTimeout(timeoutHandle);
    });
}

// Factory function to create native Promise with optional timeout
function createPromise(executor, timeout) {
    var promise = new Promise(executor);
    return (timeout > 0) ? applyTimeoutToPromise(promise, timeout) : promise;
}

// Usage

function asyncFunctionWithPromiseFactory() {
    return createPromise(function (resolve, reject) {
        // Promise executor body
    }, 5000);
}

function asyncFunctionWithNativePromise() {
    var result = new Promise(function (resolve, reject) {
        // Promise executor body
    });
    return applyTimeoutToPromise(result, 5000);
}

```

**Comment by acontreras89** [comment]
Sorry to drop by unannounced, I'm just glad someone is working on this already and I wanted to say thanks!

**Comment by mhajas** [comment]
@afedotov Thank you for looking into this thoroughly, I really appreciate it! I remember we were thinking of using just:
```
var handle = setTimeout(function () {
                    promise.setError();
                }, 5000);
```

but it didn't work properly. `Promise.race` worked much better, but I can't recall what was the exact reason behind it.

I don't see a long timeout a big problem for UX as it is reached only when Keycloak is unresponsive or wrongly configured, which should not happen so often. The biggest advantage I see in usage of `race` function is that it has almost no impact when everything works correctly when comparing with `checkIfRealmAvailable` function.

I like the code you added, but I would probably don't incorporate it into `createPromise` function as we don't need this for majority of promises. The problem is mainly in `check3pCookiesSupported` which is now causing silent failure. WDYT about something [like this](https://github.com/mhajas/keycloak/commit/2e1e9c0c47386746cfc3fc0b4899e000cec77907)? 

It uses `applyTimeoutToPromise` function you wrote and changes only behavior of `check3pCookiesSupported`. The biggest disadvantage I see is that we don't know what happens, which means we don't know difference between unresponsiveness and wrong configuration because we just know, that keycloak didn't respond in the time. However, this can be enough for solving KEYCLOAK-16455 in least invasive way.

**Comment by afedotov** [comment]
> WDYT about something [like this](https://github.com/mhajas/keycloak/commit/2e1e9c0c47386746cfc3fc0b4899e000cec77907)?

Hello, @mhajas !
These changes you suggested looks good to me! If you prefer to use timeouts, let's do it.
How we can proceed? Maybe close this PR because its title and initial idea doesn't match to chosen fix behavior?
And then you can merge your changes from a new PR?

**Comment by acontreras89** [comment]
If you've decided to change the approach to reject `check3pCookiesSupported`'s promise when this fails, please consider taking over #7779 rather than starting from scratch.

**Comment by mhajas** [comment]
@afedotov It is not on me to decide what is the better approach here. I think it would be the best to start a discussion on keycloak-dev mailing list that present both approaches. I will do so when we create a new PR. Most of the changes in the commit I sent originates from you, so if you want, you can create a new PR yourself, but I can do it if you can't, just let me know.

@acontreras89 Thanks for the heads-up. It seems that the changes in PR 7779 are very similar to these in this one. Do you mean some specific change that should be taken over?

**Comment by afedotov** [comment]
@mhajas You already have desired changes in your fork branch, so it is more handy if you make this PR.

**Comment by mhajas** [comment]
ok @afedotov I sent the PR: https://github.com/keycloak/keycloak/pull/8161. Also, I started a discussion on keycloak-dev mailing list: https://groups.google.com/g/keycloak-dev/c/GqRMAoH6OpA feel free to express your opinion there.

**Comment by acontreras89** [comment]
@mhajas I just like the approach of using `iframe.onload` rather than a timeout, mainly because it gets the job done without introducing _new elements_ to the equation.

I have been playing around with it locally and it works wonders: `onload` callback is not triggered in the happy path.

Of course this should be thoroughly tested in case it needs some tweaks (like delaying the promise cancelation to the next tick via `setTimeout(..., 0)`) for cross-browser consistency. IMHO this would still be neater than polluting the adapter with new settings and/or auxiliary functions, and it could be released as a patch rather than a minor version.

cc/ @afedotov

**Comment by afedotov** [comment]
No, @acontreras89, solution from #7779 with `iframe.onload` is bad at least because it is based on **unspecified** behavior that easily can vary/change from browser to browser. Which browsers/platforms did you tried to test? Are these tests was setup with Cross-Origin requests and TLS enabled? Is there some network latency added?

My experiments with such usage of `iframe.onload` showed that:

#### Google Chrome Version 91.0.4472.106 (macOS x86_64)

- Somehow works

#### Mozilla Firefox 89.0 (macOS x86_64)

 - In case of any network error (Connection Refused, Destination Unreachable, etc) `iframe.onload` is never being called.
 
#### Apple Safari 14.1.1 (macOS x86_64)

 - `iframe.onload` constantly being called **before** message (sent by sendMessage) arrives from iframe, sometimes TWICE in a row (look at `check3pCookiesSupported`, there are two sequential page loads in iframe, step1.html redirects to step2.html). Surprisingly the trick with setTimeout for next-tick schedule also does not work here.

 - Also in case of any network error `iframe.onload` is never being called

So at least for 2 of 3 mainstream browsers I have at hand this solution does not work.

**Comment by jonkoops** [comment]
It's quite a same that there is just no simple `onerror` event to handle this, but there is nothing we can do about it. I've also done some testing on #7779 in different browsers, and I have to agree with @afedotov's conclusion as well.

I personally think the best approach would be to handle this with a lengthy timeout (perhaps around 10s) to deal with poor network conditions.

**Comment by stianst** [comment]
This is was resolved in https://github.com/keycloak/keycloak/pull/8161 by introducing the suggested timeout, which removes the need for extra network requests. Closing this, thanks.

**Pull request closed without merging**
